packages/x-data-grid/src/hooks/features/rowSelection/useGridRowSelection.ts

```diff
@@ -446,14 +446,15 @@ export const useGridRowSelection = (
       const shouldLimitSelectionToCurrentPage =
         props.checkboxSelectionVisibleOnly && props.pagination;
 
-      const rowsToBeSelected = shouldLimitSelectionToCurrentPage
-        ? gridPaginatedVisibleSortedGridRowIdsSelector(apiRef)
-        : gridExpandedSortedRowIdsSelector(apiRef);
+      const rowsToBeSelected =
+        shouldLimitSelectionToCurrentPage && props.paginationMode === 'client'
```

Comment:
should `props.paginationMode === 'client'` be part of the `shouldLimitSelectionToCurrentPage` evaluation?

Reply:
Sure we could. Although it's a very trivial thing it seemed to me more logical to keep them separate.

`shouldLimitSelectionToCurrentPage` relates to the status `checkboxSelectionVisibleOnly`, two conditions much be satisfied in order for the `gridPaginatedVisibleSortedGridRowIdsSelector` to be used in this use case.
1. `shouldLimitSelectionToCurrentPage` is true.
2. Pagination is not "server-side"

We could even remove the variable `shouldLimitSelectionToCurrentPage` and combine the condition though.